File: frontend/mobile/lib/service/recommend_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'storage_helper.dart';
import 'auth_service.dart';
import '../config/api_config.dart';

// Service สำหรับดึงคำแนะนำอาหารและกีฬา
class RecommendationService {
  static String get baseUrl => ApiConfig.flaskUrl;

  // ดึงคำแนะนำอาหาร
  static Future<List<Map<String, dynamic>>> getFoodRecommendations({
    required int userId,
    String? date,
    int topN = 3,
  }) async {
    try {
      // ดึง access token จาก storage
      String? accessToken = await StorageHelper.getAccessToken();

      if (accessToken == null) {
        throw Exception('No access token found. Please login.');
      }

      final uri = Uri.parse('$baseUrl/api/food-recommend/$userId').replace(
        queryParameters: {
          if (date != null) 'date': date,
          'top_n': topN.toString(),
        },
      );

      var response = await http.get(
        uri,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $accessToken',
        },
      );

      // ถ้า token หมดอายุ ให้ refresh แล้วลองใหม่
      if (response.statusCode == 401 || response.statusCode == 403) {
        accessToken = await AuthService.refreshAccessToken();
        response = await http.get(
          uri,
          headers: {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer $accessToken',
          },
        );
      }

      if (response.statusCode == 200) {
        final data = json.decode(response.body);

        if (data is Map && data.containsKey('recommendations')) {
          final recs = data['recommendations'];
          if (recs is List) {
            return _convertToMapList(recs);
          }
        }

        if (data is List) {
          return _convertToMapList(data);
        }

        return [];
      } else {
        throw Exception(_parseError(response));
      }
    } catch (e) {
      rethrow;
    }
  }

  // ดึงคำแนะนำกีฬา
  static Future<List<Map<String, dynamic>>> getSportRecommendations({
    required int userId,
    int topN = 3,
    int kNeighbors = 5,
  }) async {
    try {
      // ดึง access token จาก storage
      String? accessToken = await StorageHelper.getAccessToken();

      if (accessToken == null) {
        throw Exception('No access token found. Please login.');
      }

      final uri = Uri.parse('$baseUrl/api/sport-recommend/$userId').replace(
        queryParameters: {
          'top_n': topN.toString(),
          'k_neighbors': kNeighbors.toString(),
        },
      );

      var response = await http.get(
        uri,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $accessToken',
        },
      );

      // ถ้า token หมดอายุ ให้ refresh แล้วลองใหม่
      if (response.statusCode == 401 || response.statusCode == 403) {
        accessToken = await AuthService.refreshAccessToken();
        response = await http.get(
          uri,
          headers: {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer $accessToken',
          },
        );
      }

      if (response.statusCode == 200) {
        final data = json.decode(response.body);

        if (data is Map && data.containsKey('recommendations')) {
          final recs = data['recommendations'];
          if (recs is List) {
            return _convertToMapList(recs);
          }
        }

        if (data is List) {
          return _convertToMapList(data);
        }

        return [];
      } else {
        throw Exception(_parseError(response));
      }
    } catch (e) {
      rethrow;
    }
  }

  // แปลง List ให้เป็น List<Map<String, dynamic>>
  static List<Map<String, dynamic>> _convertToMapList(List data) {
    return data.map((item) {
      if (item is Map<String, dynamic>) {
        return item;
      } else if (item is String) {
        return {
          'id': 0,
          'name': item,
          'calories': 0,
        };
      } else {
        return {
          'id': 0,
          'name': item.toString(),
          'calories': 0,
        };
      }
    }).toList();
  }

  // แปลง error message จาก response
  static String _parseError(http.Response response) {
    try {
      final body = json.decode(response.body);
      return body['message'] ?? body['error'] ?? 'Unknown error';
    } catch (_) {
      return 'Request failed (${response.statusCode})';
    }
  }
}

File: frontend/web/src/services/recommend_service.ts
import axios from 'axios';
import { getFlaskApiUrl } from '@/config/api.config';

const FLASK_API_URL = getFlaskApiUrl();

// ประเภทข้อมูลสำหรับ Recommendation
export interface FoodRecommendation {
  food_id: number;
  name: string;
  calories: number;
  similarity_score: number;
}

export interface FoodRecommendResponse {
  success: boolean;
  message: string;
  user_history: string[];
  remaining_calories: number;
  recommendations: FoodRecommendation[];
}

export interface SportRecommendResponse {
  success: boolean;
  message?: string;
  user_id?: number;
  recommendations?: string[];
  timestamp?: string;
  error?: string;
}

// สร้าง Axios Instance สำหรับ Flask API
const flaskAPI = axios.create({
  baseURL: FLASK_API_URL,
  headers: {
    'Content-Type': 'application/json',
  },
  timeout: 10000,
});

// Request Interceptor - เพิ่ม token ทุก request
flaskAPI.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('accessToken');
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

// Response Interceptor - จัดการ error และ token หมดอายุ
flaskAPI.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response?.status === 401 || error.response?.status === 403) {
      localStorage.removeItem('accessToken');
      localStorage.removeItem('user');

      if (typeof window !== 'undefined') {
        window.location.href = '/login';
      }
    }

    return Promise.reject(error);
  }
);

// Recommendation Service
export const recommendAPI = {
  // ดึงคำแนะนำอาหาร
  getFoodRecommendations: async (
    userId: number,
    topN: number = 3,
    date?: string
  ): Promise<FoodRecommendResponse> => {
    try {
      const params = new URLSearchParams();
      params.append('top_n', topN.toString());
      if (date) {
        params.append('date', date);
      }

      const response = await flaskAPI.get<FoodRecommendResponse>(
        `/api/food-recommend/${userId}?${params.toString()}`
      );

      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404 && error.response?.data) {
        return error.response.data;
      }

      const errorMessage = error.response?.data?.message || 'เกิดข้อผิดพลาดในการดึงข้อมูลแนะนำอาหาร';
      throw new Error(errorMessage);
    }
  },

  // ดึงคำแนะนำกีฬา
  getSportRecommendations: async (
    userId: number,
    topN: number = 3,
    kNeighbors: number = 5
  ): Promise<SportRecommendResponse> => {
    try {
      const params = new URLSearchParams();
      params.append('top_n', topN.toString());
      params.append('k_neighbors', kNeighbors.toString());

      const response = await flaskAPI.get<SportRecommendResponse>(
        `/api/sport-recommend/${userId}?${params.toString()}`
      );

      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404 && error.response?.data) {
        return error.response.data;
      }

      const errorMessage = error.response?.data?.message || 'เกิดข้อผิดพลาดในการดึงข้อมูลแนะนำกีฬา';
      throw new Error(errorMessage);
    }
  }
};

export default recommendAPI;
